### PR TITLE
Improve support for keyboard navigation and prevent empty server name in "Database Select" dialog.

### PR DESCRIPTION
**Pull request by AndrewDP23**
Modified both Default Database drop-downs on "Database Select" dialog to reload on enter instead of on drop-down to improve keyboard navigation.
Enabled auto-complete on both Default Database drop-downs on "Database Select" dialog to improve keyboard navigation.
Modified tab order on Form1, DatabaseSelect, Options and ParamWindow to improve keyboard navigation.
Modified "Database Select" dialog to disable OK button when Server Name is empty.

### CODE REVIEW

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 1]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+            server_textBox.TextChanged += Server_textBox_TextChanged;
+
+            Server_textBox_TextChanged(null, null);
         }
 
-        private void authentication_comboBox_SelectedIndexChanged(object sender, EventArgs e)
+        private void Server_textBox_TextChanged(object sender, EventArgs e)
         {
-            if (authentication_comboBox.SelectedIndex == 0)
+            if (server_textBox.Text == string.Empty)
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: !string.IsNullOrEmpty

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 2]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+            db_comboBox.SelectedIndex = -1;
         }
 
-        private void db_comboBox_Click(object sender, EventArgs e)
+        private void Db_comboBox_Leave(object sender, EventArgs e)
         {
-            SaveLocalSettings();
-
-            var selectedItem = (string) db_comboBox.SelectedItem;
+            ComboBox cbSender = ((ComboBox)sender);
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: Var

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 3]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+            if ((cbSender.SelectedValue == null || cbSender.Text != cbSender.SelectedItem.ToString()) && cbSender.Items.Contains(cbSender.Text))
+            {
+                cbSender.SelectedItem = cbSender.Text;
+            }
+        }
 
-            var sql = "" + "SELECT name " + "FROM master..sysdatabases " + "ORDER BY name";
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: Why concatenate - and check state also: https://github.com/ErikEJ/SqlCeToolbox/blob/master/src/GUI/SSMSToolbox/SSMSEngine/ObjectExplorerManager.cs#L72

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 4]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+        private void Db_comboBox_Enter(object sender, EventArgs e)
+        {
+            ComboBox cbSender = ((ComboBox)sender);
+            string _prevSelectedValue = cbSender.SelectedValue != null ? cbSender.SelectedValue.ToString() : string.Empty;
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: var

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 5]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+            if ((cbSender.SelectedValue == null || cbSender.Text != cbSender.SelectedItem.ToString()) && cbSender.Items.Contains(cbSender.Text))
+            {
+                cbSender.SelectedItem = cbSender.Text;
+            }
+        }
 
-            var sql = "" + "SELECT name " + "FROM master..sysdatabases " + "ORDER BY name";
+        private void Db_comboBox_Enter(object sender, EventArgs e)
+        {
+            ComboBox cbSender = ((ComboBox)sender);
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: var

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 6]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
             {
-                var comm = new SqlCommand(sql, conn);
+                cbSender.SelectedItem = _prevSelectedValue;
+            }
+        }
 
-                var databases = new List<string>();
+        private void ReloadDatabaseList(object objDatabaseParam)
+        {
+            ComboBox dbComboboxParam = (ComboBox)objDatabaseParam;
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: var

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 7]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+            string selectedComboBoxItem = (string)dbComboboxParam.SelectedItem;
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: var

File: src/SQLQueryStress/DatabaseSelect.cs [hunk 8]
```diff
@@ -84,75 +84,146 @@ public DatabaseSelect(Form1.QueryStressSettings settings)
...
+                        {
+                            databases.Add((string)reader[0]);
+                        }
                     }
-                }
-                catch (SqlException ex)
-                {
+                    catch (SqlException ex)
+                    {
                     if (ex.Number != 4060)
```
Comment (ErikEJ) [inline src/SQLQueryStress/DatabaseSelect.cs]: Wonder why??

File: src/SQLQueryStress/DatabaseSelect.cs (hunk 1 above)
Comment (AndrewDP23) [inline src/SQLQueryStress/DatabaseSelect.cs]: That's much more elegant than what I've done here.

File: src/SQLQueryStress/DatabaseSelect.cs (hunk 8 above)
Comment (AndrewDP23) [inline src/SQLQueryStress/DatabaseSelect.cs]: The two database drop-downs (under "Main Load Settings" and "Parameterization Settings") are being reloaded using very similar procedures, and I saw a slight improvement in code maintainability by merging the two procedures.  I will separate this into two separate procedures if requested.

File: src/SQLQueryStress/DatabaseSelect.cs (hunk 3 above)
Comment (AndrewDP23) [inline src/SQLQueryStress/DatabaseSelect.cs]: I'm not sure why this line was fragmented. I suspect a previous developer may have tried to keep each clause of this SELECT statement on separate lines. I've tidied this and added a state filter to a commit for which I'll submit another pull request in a moment.

**Pull request merged**